### ISSUE TITLE
manual: "tail-module-cons" motivating example fails to fail with a stackoverflow

### ISSUE THREAD

**Issue by Octachron** [ocaml/ocaml#12000]
With the OCaml 5 runtime, the introductory example of the tail-modulo-cons chapter 
(see https://ocaml.org/releases/5.0/manual/tail_mod_cons.html for a rendered version):

>```ocaml
>let rec map f l =
>  match l with
>  | [] -> []
>  | x :: xs ->
>    let y = f x in
>    y :: map f xs
>```
>...
>This is a correctness issue for large lists on operating systems with limited stack space – the dreaded Stack_overflow exception.
>```ocaml
># List.length (map Fun.id (List.init 1_000_000 Fun.id));;
>- : int = 1000000
>```
fails to fail with a `Stack_overflow` exception due to the far less limited stack size of the new runtime.

I am wondering how the text should be updated: Should we remove the `Stack_overflow` example? Should we run the examples with a limited stack size? Should we mention that this was a problem for previous versions of OCaml?

**Comment by gasche** [comment]
For the tutorial, does it work to just use `1000_000_000`?

In the long run, the question would be: when/why are we advocating for TMC now?
1. For 4.x it is clearly about correctness. For 5.x it is less clear: you can make the OCaml stack as large as you want now, but this creates other usability issues (it makes some non-termination bugs much more painful).
2. For 5.x we still get an efficiency benefit on large inputs (because trashing the call stack has poor locality).

We can still explain the benefits, but the story is more complex than before. (This is good news, it means that 5.x will work better for many programs without requiring the use of TMC as an advanced feature.)

**Comment by hhugo** [comment]
> For the tutorial, does it work to just use `1000_000_000`?
> 
> In the long run, the question would be: when/why are we advocating for TMC now?
> 
> 1. For 4.x it is clearly about correctness. For 5.x it is less clear: you can make the OCaml stack as large as you want now, but this creates other usability issues (it makes some non-termination bugs much more painful).
> 2. For 5.x we still get an efficiency benefit on large inputs (because trashing the call stack has poor locality).
> 
> We can still explain the benefits, but the story is more complex than before. (This is good news, it means that 5.x will work better for many programs without requiring the use of TMC as an advanced feature.)

TMC is still useful for jsoo where the stack is much smaller (when the code is not cps transformed)

**Comment by Octachron** [comment]
Using `1_000_000_000` for the list size has the issue that the initial list takes nearly one minute to build on my machine.
We could move closer to the edge at `25_000_000` which does produce a stack overflow in 3 seconds (the current example takes 0.5 second on this machine).

Another option might be to use a cyclic list (which decreases the evaluation time to 1.5 seconds and ensure that we will get a stack overflow ... or a worse issue) ?

**Comment by xavierleroy** [comment]
I spent time fixing the test suite around this issue.  If you want to reliably and quickly trigger a stack overflow, just run with `OCAMLRUNPARAM=l=100k` or something like this.  

BTW, https://v2.ocaml.org/manual/runtime.html may need fixing: "This is only relevant to the byte-code runtime, as the native code runtime uses the operating system’s stack. " no longer holds in 5.0

**Comment by Octachron** [comment]
> If you want to reliably and quickly trigger a stack overflow, just run with OCAMLRUNPARAM=l=100k or something like this.

I thought about this option, but this doesn't work that well in the manual when we want the example to be easily reproducible by an user copy-pasting the example in their toplevel which is unlikely to have been launched with any `OCAMLRUNPARAM`s.

**Comment by yawaramin** [comment]
Isn't a big motivation to allow running in constant memory? Non-tail recursive functions will keep growing memory usage until they finish, right?

**Comment by xavierleroy** [comment]
> we want the example to be easily reproducible by an user copy-pasting the example in their toplevel

Then you could include in the source code something like `Gc.(set {get() with stack_limit = ...})`.  (Not sure it works in OCaml 5.)  But that's a lot of noise for a silly example.  Generally speaking, we want to show good code to users, not code that is known to blow up.

**Comment by Octachron** [comment]
I agree that is a good idea to be very explicit (with a visible call to `Gc.set`) on the fact that we are cornering the runtime into a bad behavior.

@yawaramin for TMC to be useful, the TMC call needs to be allocating a new data structure. Allocating a new block (with TMC) instead of allocating a new block *and* a new frame (without TMC) can only change a constant factor in the space complexity.

**Comment by gasche** [comment]
I cannot observe a Stack_overflow under the 5.0 toplevel after setting `stack_limit` using `Gc.set`.

```ocaml
# let with_stack_limit stack_limit f =
    let old_gc_settings = Gc.get () in
    Gc.set { old_gc_settings with stack_limit };
    Fun.protect ~finally:(fun () -> Gc.set old_gc_settings) f
  ;;
val with_stack_limit : int -> (unit -> 'a) -> 'a = <fun>

(* check that with_stack_limit works *)
# (with_stack_limit 20_000 Gc.get).stack_limit;;
- : int = 20000

(* non-tail-rec map function *)
# let rec map f = function
  | [] -> []
  | x :: xs -> f x :: map f xs;;
val map : ('a -> 'b) -> 'a list -> 'b list = <fun>

# let large_list = List.init 1_000_000 Fun.id;;
val large_list : int list = [...]

(* this should fail with a Stack Overflow *)
# with_stack_limit 20_000 (fun () -> map Fun.id large_list |> List.length);;
- : int = 1000000
```

On the other hand, running `OCAMLRUNPARAM="l=20000" ocaml` does work.

It looks like there is a bug with the `stack_limit` Gc control.

**Comment by gasche** [comment]
Ah, in fact this works properly in trunk, I believe that it was fixed by #11054. Sorry for the noise.

**Comment by gasche** [comment]
I propose a fix using `Gc.set` in #12077. Reviews welcome.
